torch/_inductor/runtime/triton_helpers.py

```diff
@@ -384,3 +379,162 @@ def frexp(x):
+        cond = cond | ((left == right) & (left_idx > right_idx))
+
+    cond = (right_valid_mask > left_valid_mask) | (
+        (right_valid_mask == left_valid_mask) & cond
+    )
+    cond = cond ^ flip
+    ret = ix ^ tl.where(cond, ileft ^ iright, tl.zeros_like(ix))
+    new_idxs = idxs ^ tl.where(cond, left_idx ^ right_idx, tl.zeros_like(idxs))
+    if valid_mask is None:
+        new_valid_mask = tl.full(x.shape, True, tl.int1)
+    else:
+        new_valid_mask = valid_mask ^ tl.where(
+            cond, left_valid_mask ^ right_valid_mask, tl.zeros_like(valid_mask)
+        )
+
+    return ret.to(x.dtype, bitcast=True), new_idxs, new_valid_mask
+
+
+@triton.jit
+def _bitonic_merge_with_index(
+    x,
+    idxs,
+    mask,
+    stage: tl.constexpr,
+    alternating: tl.constexpr,
+    n_dims: tl.constexpr,
+    stable: tl.constexpr,
+    descending: tl.constexpr,
+):
+    """
+    order_type 0 == ascending
+    order_type 1 == descending
+    order_type 2 == alternating
+    """
+    n_outer: tl.constexpr = x.numel >> n_dims
+    tl.static_assert(stage <= n_dims)
+    # flip denotes whether to re-arrange sub-sequences of elements in ascending or
+    # descending order.
+    # if flip = 00000000... then all elements will be re-arranged ascendingly at this stage
+    # if flip = 00110011... then all the elements will be re-arranged alternatingly (with
+    # a stride of 2) at this stage
+    if alternating:
+        shape: tl.constexpr = [n_outer * 2 ** (n_dims - 1 - stage), 2, 2**stage]
+        flip = tl.reshape(
+            tl.broadcast_to(tl.arange(0, 2)[None, :, None], shape), x.shape
+        )
+    else:
+        flip = False
+    # perform `stage` rounds of `compare-and-swap`
+    next_mask = mask
+    for i in tl.static_range(stage):
+        x, idxs, next_mask = _compare_and_swap_with_index(
+            x, idxs, mask, flip, i + (n_dims - stage), n_dims, stable, descending
+        )
+        if mask is not None:
+            mask = next_mask
+    return x, idxs, next_mask
+
+
+@triton.jit
+def sort_with_index(
+    x,  # value
+    idxs,  # index
+    mask,  # mask if current value is valid (invalid values sort to the end)
+    dim: tl.constexpr = None,
+    stable: tl.constexpr = tl.constexpr(False),
+    descending: tl.constexpr = tl.constexpr(False),
+):
+    x, idxs = tl.broadcast(x, idxs)
+    if mask is not None:
+        x, mask = tl.broadcast(x, mask)
+    # handle default dimension or check that it is the most minor dim
+    _dim: tl.constexpr = len(x.shape) - 1 if dim is None else dim
+    tl.static_assert(
+        _dim == len(x.shape) - 1, "only minor dimension is currently supported"
+    )
+    # iteratively run bitonic merge-sort steps
+    n_dims: tl.constexpr = _log2(x.shape[_dim])
+
+    for i in tl.static_range(1, n_dims + 1):
+        x, idxs, next_mask = _bitonic_merge_with_index(
+            x,
+            idxs,
+            mask,
+            i,
+            alternating=i < n_dims,
```

Comment:
Just for me to understand, is this an optimisation in the last iteration or is there more to it?

Reply:
It's just part of the algorithm. The first `n_dims` iterations are the divide and conquer part which alternate in sorting direction, then we do the final merge to get it all in the same direction at the end.

https://www.baeldung.com/cs/bitonic-sort#3-pseudocode

(I think that pseudo-code is for a slightly different sorting network, but same idea)